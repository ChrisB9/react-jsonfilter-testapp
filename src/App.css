.App {
}

.App-logo {
  height: 30px;
  background: white;
  border-radius: 50%;
  padding: 10px;
  border: 3px solid #b465da;
}

.App-header {
  background-color: #222;
  color: white;
}

.App-intro {
  font-size: large;
  text-align: center;
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: column;
  margin: 1% 10%;
}

input {
	padding: .5em .6em;
    display: inline-block;
    border: 1px solid #ccc;
    box-shadow: inset 0 1px 3px #ddd;
    border-radius: 4px;
    vertical-align: middle;
    box-sizing: border-box;
}

table {
  padding: 0;
  border-collapse: collapse;
}
table tr {
	background-color: white;
	margin: 0;
	padding: 0;
}
table tr:nth-child(2n) {
  background-color: #f8f8f8;
}
table tr th {
  font-weight: bold;
  border-bottom: 1px solid #cccccc;
  text-align: left;
  margin: 0;
  padding: 6px 13px;
}
table tr td {
  text-align: left;
  margin: 0;
  padding: 6px 13px;
}
table tr th :first-child, table tr td :first-child {
  margin-top: 0;
}
table tr th :last-child, table tr td :last-child {
  margin-bottom: 0;
}