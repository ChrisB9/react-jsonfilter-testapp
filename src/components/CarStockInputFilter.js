import React from 'react';

const CarStockInputFilter = (props) => {
	return (
		<input onChange={props.filter} placeholder="type something to filter table" />
	)
}

export default CarStockInputFilter;
