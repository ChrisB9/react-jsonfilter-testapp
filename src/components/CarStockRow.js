import React from 'react';

const CarStockRow = (props) => {
	return (
		<tr>
			{props.properties.map((prop, index) => <td key={index}>{props.data[prop]}</td>)}
		</tr>
	)
}

export default CarStockRow;