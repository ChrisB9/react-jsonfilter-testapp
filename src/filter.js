
const filter = (data, input) => {
  let newArray = []
  data.map((car, key) => {
    let objKeys= Object.keys(car);
    for (let i = 0; i < objKeys.length; i++) {
      if (input.includes((objKeys[i]+":").toLowerCase())) {
        let str = input.replace((objKeys[i]+": ").toLowerCase(), "");
        return (""+car[objKeys[i]]).toLowerCase().includes(str) ? key : -1;
      }
    }
    return typeof Object.keys(car).map((key) => car[key]).find((el) => {
      return (""+el).toLowerCase().includes(input)
    }) === 'undefined' ? -1 : key;
  }).forEach((index) => {
    if (index !== -1)
      newArray.push(data[index])
  })
  return newArray
}

export default filter;