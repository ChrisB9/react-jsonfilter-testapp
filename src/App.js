import React, { Component } from 'react';
import CarStockTable from './components/CarStockTable';
import CarStockInputFilter from './components/CarStockInputFilter';
import filter from './filter';
import logo from '../public/favicon-96x96.png';
import './App.css';
import data from './data.json'

class App extends Component {
  constructor(props) {
    super(props);
  
    this.state = {
      carStock : this.addIdToObject(data.carStock)
    };
    this.eventHandler = this.eventHandler.bind(this);
  }

  addIdToObject(object) {
    return object.map((el, index) => {
      el.id = ++index;
      return el
    })
  }

  eventHandler(event) {
    this.setState({
      carStock: filter(data.carStock, event.target.value.toLowerCase())
    })
  }

  keysFromObject(object) {
    let set = new Set();
    object.forEach((el) => {
      Object.keys(el).forEach((property) => {set.add(property)})
    });
    let arr = Array.from(set);
    arr.splice(arr.indexOf("id"), 1);
    return arr;
  }

  componentDidUpdate() {
    console.log("app updated")
  }

  render() {
    return (
      <div className="App">
        <div className="App-header">
          <img src={logo} className="App-logo" alt="logo" />
          car stock filter
        </div>
        <div className="App-intro">
          <CarStockInputFilter filter={this.eventHandler}/>
          <CarStockTable 
            head={["id", ...this.keysFromObject(data.carStock)]} 
            carStock={this.state.carStock} />
        </div>
      </div>
    );
  }
}

export default App;
