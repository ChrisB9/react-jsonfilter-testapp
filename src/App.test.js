import React from 'react';
import ReactDOM from 'react-dom';
import ReactTestUtils from 'jest';
import { expect as chaiExpect, assert } from 'chai';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import App from './App';
import CarStockRow from './components/CarStockRow'
import CarStockInputFilter from './components/CarStockInputFilter'
import data from './data.json';
import filter from './filter';

describe('App component', () => {
  	it('renders without crashing', () => {
		const div = document.createElement('div');
		ReactDOM.render(<App />, div);
	});
	it('match jest tree', () => {
		const component = renderer.create(<App />);
		let tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});

	it('should contain a table', () => {
		const app = shallow(<App />);
		expect(app.find('table')).toBeDefined();
	});
});

describe('CarStockRow component', () => {
	it('constains tr element with id', () => {
		const component = shallow(
			<CarStockRow 
			data={{id: 1}}
			properties={["id"]} />
		);
		chaiExpect(component.find('tr').children()).to.have.length(1)
	});
});

describe('CarStockInput component', () => {
	let test;
	const mockEventFunction = (event) => {test = event.target.value};
	it('should trigger on change', () => {
		const component = shallow(
			<CarStockInputFilter 
			filter={mockEventFunction} />
		);
		component.find('input').simulate('change', {target: {value: 'My new value'}});
		chaiExpect(test).to.equal('My new value');
	});
});

describe('filter function tests', () => {
	it('should return an array', () => {
		assert.isArray(filter(data.carStock, "red"));
	});

	it('should have a price, color and note key', () => {
		chaiExpect(filter(data.carStock, "8000")[0]).to.have.any.key('price', 'color', 'note');
	});

	it('should have 4 elements by searching for "19"', () => {
		chaiExpect(filter(data.carStock, "19")).to.have.length(4);
	});

	it('should contain a single complete result searching with tag "id: 6"', () => {
		chaiExpect(filter(data.carStock, "id: 6")).to.eql([{
			"color": "gray",
            "type": "Audi Commodore",
            "yearOfConstruction": 1992,
            "price": 8212,
            "id": 6
		}]);
	});

	it('should have a note searching for "dam"', () => {
		chaiExpect(filter(data.carStock, "dam")[0]).to.contain.key('notes');
	});
});

