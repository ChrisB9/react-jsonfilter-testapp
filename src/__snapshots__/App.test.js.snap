exports[`App component match jest tree 1`] = `
<div
  className="App">
  <div
    className="App-header">
    <img
      alt="logo"
      className="App-logo"
      src="favicon-96x96.png" />
    car stock filter
  </div>
  <div
    className="App-intro">
    <input
      onChange={[Function]}
      placeholder="type something to filter table" />
    <table>
      <thead>
        <tr>
          <th>
            id
          </th>
          <th>
            color
          </th>
          <th>
            price
          </th>
          <th>
            type
          </th>
          <th>
            yearOfConstruction
          </th>
          <th>
            notes
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            1
          </td>
          <td>
            red
          </td>
          <td>
            60000
          </td>
          <td>
            Porsche 911
          </td>
          <td>
            2000
          </td>
          <td />
        </tr>
        <tr>
          <td>
            2
          </td>
          <td>
            black
          </td>
          <td>
            120000
          </td>
          <td>
            Porsche Panamera
          </td>
          <td>
            2010
          </td>
          <td />
        </tr>
        <tr>
          <td>
            3
          </td>
          <td>
            green
          </td>
          <td>
            8000
          </td>
          <td>
            VW Beetle
          </td>
          <td>
            1960
          </td>
          <td>
            damaged at the front bumper
          </td>
        </tr>
        <tr>
          <td>
            4
          </td>
          <td>
            blue
          </td>
          <td>
            28345
          </td>
          <td>
            Audi A5
          </td>
          <td>
            1998
          </td>
          <td />
        </tr>
        <tr>
          <td>
            5
          </td>
          <td>
            yellow
          </td>
          <td>
            80435
          </td>
          <td>
            Ferrari 430 Spider
          </td>
          <td>
            1990
          </td>
          <td />
        </tr>
        <tr>
          <td>
            6
          </td>
          <td>
            gray
          </td>
          <td>
            8212
          </td>
          <td>
            Audi Commodore
          </td>
          <td>
            1992
          </td>
          <td />
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
